File: scripts/GroupSubscriber.py
#!/usr/bin/python
import os
import rospy
from openhab_msgs.msg import GroupState

class GroupSubscriber(object):

    def __init__(self, item_name):
        self.item_name = item_name

        """Configure subscriber."""
        # Create a subscriber with appropriate topic, custom message and name of
        # callback function.
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", GroupState, self.callback)

        # Initialize message variables.
        self.enable = False
        self.data = ""

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        self.enable = True
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", GroupState, self.callback)

    def callback(self, data):
        """Handle subscriber data."""
        # Simply print out values in our custom message.
        self.data = data
        if self.data.isbool == False:
            msg = "Received %s" % (self.data.item)
        else:
            msg = "Received %s with NULL" % self.data.item
        rospy.loginfo(rospy.get_caller_id() + msg)

    def stop(self):
        """Turn off subscriber."""
        self.enable = False
        self.sub.unregister()

if __name__ == '__main__':
    # Initialize the node and name it.
    node_name = "GroupSubscriberNode"
    rospy.init_node(node_name, anonymous=True)

    groupSubscriber = GroupSubscriber("Static")

    # Go to the main loop
    try:
        groupSubscriber.start()
        # Wait for messages on topic, go to callback function when new messages arrive.
        # spin() simply keeps python from exiting until this node is stopped
        rospy.spin()
    # Stop with Ctrl + C
    except KeyboardInterrupt:
        groupSubscriber.stop()

        nodes = os.popen("rosnode list").readlines()
        for i in range(len(nodes)):
            nodes[i] = nodes[i].replace("\n","")

        for node in nodes:
            os.system("rosnode kill " + node_name)


        print("Node stopped")

File: src/GroupSubscriber.cpp
#include <ros/ros.h>
#include <openhab_msgs/GroupState.h>

class GroupSubscriber {
public:
    GroupSubscriber(std::string item_name) : item_name_(item_name), enable_(false) {
        // Create a subscriber with appropriate topic, custom message and name of callback function.
        sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 1, &GroupSubscriber::callback, this);
    }

    void start() {
        enable_ = true;
        sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 1, &GroupSubscriber::callback, this);
    }

    void callback(const openhab_msgs::GroupState::ConstPtr& data) {
        // Simply print out values in our custom message.
        data_ = data;
        std::string msg;
        if (!data_->isnull) {
            msg = "Received " + data_->item;
        } else {
            msg = "Received " + data_->item + " with NULL";
        }
        ROS_INFO_STREAM(ros::this_node::getName() << ": " << msg);
    }

    void stop() {
        enable_ = false;
        sub_.shutdown();
    }

private:
    std::string item_name_;
    bool enable_;
    ros::NodeHandle nh_;
    ros::Subscriber sub_;
    openhab_msgs::GroupState::ConstPtr data_;
};

int main(int argc, char** argv) {
    // Initialize the node and name it.
    std::string node_name = "GroupSubscriberNode";
    ros::init(argc, argv, node_name);
    ros::NodeHandle nh;

    GroupSubscriber groupSubscriber("Static");

    // Go to the main loop
    try {
        groupSubscriber.start();
        // Wait for messages on topic, go to callback function when new messages arrive.
        // spin() simply keeps C++ from exiting until this node is stopped
        ros::spin();
    } catch (const ros::Exception& e) {
        groupSubscriber.stop();

        std::vector<std::string> nodes;
        ros::master::getNodes(nodes);
        for (const auto& node : nodes) {
            system(("rosnode kill " + node_name).c_str());
        }

        ROS_INFO_STREAM(ros::this_node::getName() << ": Node stopped");
    }

    return 0;
}
